src/java/org/apache/cassandra/transport/messages/StartupMessage.java

```diff
@@ -118,8 +121,29 @@ else if (compression.equals("lz4"))
             clientState.setDriverVersion(options.get(DRIVER_VERSION));
         }
 
-        if (DatabaseDescriptor.getAuthenticator().requireAuthentication())
+        IAuthenticator authenticator = DatabaseDescriptor.getAuthenticator();
+        if (authenticator.requireAuthentication()) {
+            // If the authenticator supports early certificate authentication, attempt to authenticate with certificates.
+            if (authenticator.supportsEarlyCertificateAuthentication()) {
+                IAuthenticator.SaslNegotiator negotiator = ((ServerConnection) connection).getSaslNegotiator(state);
+                // If the negotiator determines that certificate authentication is required, attempt to authenticate on it.
+                if (negotiator.requiresCertificateAuthentication()) {
+                    // Attempt to authenticate the user.
+                    return AuthUtil.handleLogin(connection, state, EMPTY_CLIENT_RESPONSE, (negotiationComplete, challenge) ->
+                    {
+                        if (negotiationComplete) {
+                            // Authentication was successful, proceed.
+                            return new ReadyMessage();
+                        } else {
+                            // It's expected that any negotiator that requires a challenge will likely not support early
+                            // certificate authentication, in this case we can just go through the traditional auth flow.
+                            return new AuthenticateMessage(DatabaseDescriptor.getAuthenticator().getClass().getName());
```

Comment:
Just occurred to me - if we're in the "fallback" part of a client-cert-fallback-to-credentials authenticator, shouldn't the client not need to know that it's a custom authenticator, and just receive an AUTHENTICATE with PasswordAuthenticator? Otherwise, we'd have to make sure all clients are aware of a new authenticator.

Doesn't seem like a new issue with this code, just wondering how we handle this.

Reply:
Interesting - might cause problems in the future (or with niche clients), good to know. Thanks!

Reply:
yeah - although we didn't change anything here in this PR, should behave like it did before :+1:

Reply:
Yep, this was all adjacent curiosity. PR LGTM.